src/microstates.ts

```diff
@@ -0,0 +1,44 @@
+import traverseState from './utils/traverseState';
+import traverseActions from './utils/traverseActions';
+import { IClass, IMicrostate, IObserver, IState, IStateObject } from './Interfaces';
+
+export default function microstates(Class: IClass, initial: IStateObject = {}): IMicrostate {
+  if (typeof Class !== 'function') {
+    throw new Error(
+      `microstates() expects first argument to be a class, instead received ${typeof Class}`
+    );
+  }
+
+  let observer: IObserver;
+
+  let unsubscribe = (): void => (observer = null);
+
+  let subscribe = (_observer: IObserver) => {
+    observer = _observer;
+    return {
+      unsubscribe,
+    };
+  };
+
+  let onChange = (newState: IState) => {
+    if (observer) {
+      observer.next(microstate(Class, newState));
+    } else {
```

Comment:
> In other words, the transitions aren't pure functions, but instead depend on some shared mutable state. This comes from intermingling the logical transitions of a state (pure) with the tracking of that state over time (impure).

Yes.

> In other words, a microstate is data + transitions, and an observable hold the microstate and tracks transitions.

Yes!, but it'd need to be something like this:

```js
let ms = microstate(Boolean, false);

let observable = Observable.from(ms);

let { state, actions } = ms;
```

To do this, our `{ state, actions }` needs to have [`{}[Symbol.observable}`](https://github.com/tc39/proposal-observable#observablefrom) on it. This will allow ES7 Observable convert our hash into an observable.